### ISSUE TITLE
Bug: Updating a checkbox's indeterminate and checked states in the same tick causes a display issue in Chrome

### ISSUE THREAD

**Issue by lucaschen** [facebook/react#18342]
React version: 16.12.0

## Steps To Reproduce

Please see this CodeSandbox: [https://codesandbox.io/s/unruffled-mirzakhani-tb81d](https://codesandbox.io/s/unruffled-mirzakhani-tb81d)

Firstly, you'll see that checking/unchecking the checkbox will toggle the visibility of `span.indicator` - that's fine.

Secondly, please press `Do 1`, `Do 2`, `Do 3` in that order - you'll notice that the checkbox goes from indeterminate (after Do 1) to unchecked (after Do 2) to checked (Do 3) - `span.indicator` still shows correctly here and works fine.

Finally, please press `Do 1` and then `Do 3` - **this is where the bug is**. You'll notice that the checkbox is checked, but `span.indicator` does not show. If you do a further investigation into `span.indicator` you'll notice that it has `display: inline;` (i.e. should show), but for some reason it does not. The `:checked + .indicator` styles also appear to be applying inside of the DevTools as well.

## The expected behavior

`span.indicator` is meant to show after pressing `Do 1` and then `Do 3`.

## What I think the issue is

Firstly, it only happens in Chrome - works completely fine in Firefox and also Safari. No, it's not my extensions - I used a guest session and also tried on another computer - same issue.

When `indeterminate` and `checked` are updated in the same tick, then we get this issue - I'm not sure how ReactDOM changes the `checked` property on the checkbox internally (I could not reproduce this issue in Vanilla JavaScript), but the way it's doing it by updating `checked` and then `indeterminate` later on seems to be causing a weird Chrome glitch.

Please let me know if there is any additional information I can provide.

**Comment by eps1lon** [comment]
I'm fairly certain this is how your code updates the `indeterminate` property:

```js
useEffect(() => {
  checkboxRef.current.indeterminate = indeterminate;
}, [indeterminate]);
```

This means that react will commit a version of your UI with a new `indeterminate` and after that commit it will run the effect. This means there will be two different paints. If you want to set the property on the element in the same commit as react commits your UI with a new `indeterminate` you need to use [`useLayoutEffect`](https://reactjs.org/docs/hooks-reference.html#uselayouteffect)

**Comment by bvaughn** [comment]
I see the behavior you're describing when clicking 1 -> 3:
<img width="125" alt="Screen Shot 2020-03-20 at 10 02 27 AM" src="https://user-images.githubusercontent.com/29597/77187570-edc2f300-6a91-11ea-8652-55fef83f3499.png">

But Chrome's Elements panel shows the right styles are applied:
<img width="363" alt="Screen Shot 2020-03-20 at 10 02 24 AM" src="https://user-images.githubusercontent.com/29597/77187566-ed2a5c80-6a91-11ea-871f-4246b881064f.png">

This looks like a Chrome bug at first glance... if I manually toggle something on/off in the style panel, then the component renders correctly. The computed tab also shows that it should be displayed:
<img width="327" alt="Screen Shot 2020-03-20 at 10 07 13 AM" src="https://user-images.githubusercontent.com/29597/77188028-98d3ac80-6a92-11ea-932a-eff3374ebbdd.png">

**Comment by eps1lon** [comment]
It seems this is some weirdness when setting `.checked = true` while `.indeterminate === true`.

Even this vanilla html+js (which is basically what your codesandbox is doing by setting `indeterminate=true` in one render, `checked=true` in the next and in a passive effect `indeterminate=false`.

```js

document.getElementById("app").innerHTML = `
<div class="checkbox">
      <input value="1" type="checkbox" />
      <span class="indicator">checked</span>
    </div>
`;

const checkbox = document.querySelector("input");

checkbox.indeterminate = true;

setTimeout(() => {
  console.log("checked=true");
  checkbox.checked = true;

  setTimeout(() => {
    console.log("indeterminate=false");
    checkbox.indeterminate = false;
  }, 1000);
}, 1000);
```

produces different results in chrome and firefox:

chrome: visually indeterminate
firefox: visually indeterminate but `:checked` still matches leading to the checked indicator being displayed.

Maybe React could normalize this by clearing `indeterminate` when the `checked` state changes but this is pretty dangerous since there's probably code out there relying on the current behavior.

In other words: Put your state in an enum rather than spreading it across flags and rendering inconsistent combinations :wink:

**Comment by kernyden** [comment]
seems like it can be done with <code>&:checked:not(&:indeterminate)</code>
